// scrapper/lib/scrapper.js
var fs = require('fs');
var querystring = require('querystring');
var request = require('request');
var cheerio = require('cheerio');

exports.post = function (url, cb, data, headers) {
    var params = {
        url: encodeURI(url),
        method: 'POST',
        headers: {
            'User-Agent': 'Mozilla/5.0 (Windows NT 6.1) AppleWebKit/535.11 (KHTML, like Gecko) Safari/535.11',
            'Referer': url
        }};

    if (typeof data == 'object') {
        params.body = querystring.stringify(data);
    }

    if (typeof headers == 'object') {
        for (var i = 0; i < headers.length; i++) {
            params.headers[i] = headers[i];
        }
    }

    request(params, function (err, res, body) {
        if (err) return cb(err);

        exports.load(body, cb);
    });
}

exports.get = function (url, cb, data, headers) {
    if (typeof data == 'object') {
        url += (url.substr(-1) == "?") ? '' : '?';
        url += querystring.stringify(data);
    }

    var params = {
        url: encodeURI(url),
        method: 'GET',
        headers: {
            'User-Agent': 'Mozilla/5.0 (Windows NT 6.1) AppleWebKit/535.11 (KHTML, like Gecko) Safari/535.11',
            'Referer': url
        }};

    if (typeof headers == 'object') {
        for (var i = 0; i < headers.length; i++) {
            params.headers[i] = headers[i];
        }
    }

    request(params, function (err, res, body) {
        if (err) return cb(err);

        exports.load(body, cb);
    });
}

exports.load = function (html, cb) {
    var jq = cheerio.load(html);

    if (cb) cb.call(jq, jq);
}
