// hhnh-js/lib/hhnh-js.js
//load modules
var http= require('http');
var request = require('request');
var $=require('cheerio');
var zlib = require('zlib');

// module vars
var current_search_page=1;
var videos_responses = new Array();
var hhnh = {};

// search for videos
hhnh.searchSongs = function(query,page,order,cb) {
    var totalresults = 0;
    var q = query.replace(/ /g,'+');
    try{
       mixtape_responses = new Array();
        var req=http.request('http://www.hotnewhiphop.com/search/songs/'+q+'/'+order+'/'+page+'/');
        var datas;
        req.on('response',function(response) { 
            if(response.headers['content-encoding'] == 'gzip') {
                var gunzip = zlib.createGunzip();
                var json = "";
                gunzip.on('data', function(data){
                    json += data.toString();
                });
                gunzip.on('end', function(){
                    datas = json;
                    makeSongsLists(datas,cb);
                });
                response.pipe(gunzip);
            } else {
                var data = new Array(); 
                response.on("data", function(chunk) {
                    data.push(chunk);
                });
                response.on('end',function(){
                    datas = data.join('');
                    makeSongsLists(datas,cb);
                });
            }
        });
            req.on('error', function(e) {
                console.log("Got error: " + e.message);
            });
            req.end();
    } catch(err){
        console.log('hhnh topmixtapes err: '+err);
    }
}

hhnh.topMixtapes = function (filters,cb) {
    var totalResults = 0;
    try{
       mixtape_responses = new Array();
        var req=http.get('http://www.hotnewhiphop.com/mixtapes/top100/mainstream/'+filters+'/');
        req.on('response',function(response) { 
            var data = new Array(); 
            response.on("data", function(chunk) {
                data.push(chunk)
            });
            response.on('end',function(){
                var datas = data.join('');
                var list = $('.mixtape-grid-list div.newCell',datas);
                var results = {};
                results.total = list.length;
                results.cb = cb;
                results.items = [];
                results.totalResults = list.length;
                $(list).each(function(t,r){
                    var obj={};
                    var v = $(r);
                    obj.link=v.find('a.flagLink')[0].attribs['href'];
                    obj.thumb=v.find('img')[0].attribs['src'];
                    obj.title=v.find('a.flagLink')[0].attribs['title'];
                    obj.artist=v.find('i')[0].children[0].data;
                    storeVideosInfos(results,obj,t);
                });
            });
        });
        req.on('error', function(e) {
            console.log("Got error: " + e.message);
        });
        req.end();
    } catch(err){
        console.log('hhnh topmixtapes err: '+err);
    }
}

hhnh.loadMixtapeSongs = function (link,cb) {
    var totalResults = 0;
    var url = link.replace('http://www.hotnewhiphop.com','');
    var options = 
    {
        host: 'www.hotnewhiphop.com',
        headers: {'user-agent': 'Mozilla/5.0 (X11; Linux i686) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/29.0.1547.0 Safari/537.36'},
        path: url,
        encoding: null
    }
    try{
       mixtape_responses = new Array();
        var req=http.request(options);
        var datas;
        req.on('response',function(response) { 
            if(response.headers['content-encoding'] == 'gzip') {
                var gunzip = zlib.createGunzip();
                var json = "";
                gunzip.on('data', function(data){
                    json += data.toString();
                });
                gunzip.on('end', function(){
                    datas = json;
                    prepareTrack(datas,cb);
                });
                response.pipe(gunzip);
            } else {
                var data = new Array(); 
                response.on("data", function(chunk) {
                    data.push(chunk);
                });
                response.on('end',function(){
                    datas = data.join('');
                    prepareTrack(datas,cb);
                });
            }
        });
    
            req.on('error', function(e) {
                console.log("Got error: " + e.message);
            });
            req.end();
    } catch(err){
        console.log('hhnh topmixtapes err: '+err);
    }
}

function makeSongsLists(datas,cb) {
    var list = $('div.newCell',datas);
    var results = {};
    results.total = list.length;
    results.cb = cb;
    results.items = [];
    results.totalResults = list.length;
    $(list).each(function(t,r){
        var obj={};
        var v = $(r);
        obj.link=v.find('.centerBlock h3 a')[0].attribs['href'];
        obj.thumb=v.find('img')[0].attribs['src'];
        obj.title=v.find('.centerBlock h3 a')[0].attribs['title'];
        obj.artist=v.find('i')[0].children[0].data;
        console.log(results,obj,t);
        storeVideosInfos(results,obj,t);
    });
}

function prepareTrack(datas,cb) {
    try {
                var thumb = $('.coverImage1',datas)[0].attribs['src'];
                var author = $('.player-page a',datas)[0].children[0].data;
                var list = datas.match(/existingPlaylist = (.*)/)[1];
            } catch(err) {
                console.log("Got error: " + err.message, datas);
                return;
            }
            var results= {};
            var arr = list.replace('[','').replace("}];",'').split('},');
            results.cb = cb;
            results.items = [];
            results.totalResults = arr.length;
            results.total = arr.length;
            $(arr).each(function(t,r){
                var x = r+'}';
                try {
                    var track = JSON.parse(x);
                    track.thumb = thumb;
                    track.author = author;
                    track.resolutions=[];
                    track.views='...';
                    track.resolutions['360p'] = {};
                    track.resolutions['360p']['link'] = track.mp3;
                    track.resolutions['360p']['container'] = 'mp3';
                } catch(err) {
                    console.log("Got error: " + err.message);hhnh.loadMixtapeSongs(link,cb);
                }
                storeVideosInfos(results,track,t);
            });
}


// store videos and return it in the right order...
function storeVideosInfos(video,infos,num) {
    video.items.push(infos); 
    videos_responses[num]=video;
    if (videos_responses.length == video.total) {
        video.cb(videos_responses);
        videos_responses = new Array();
    }
}

module.exports = hhnh;
