// yt-streamer/lib/yt-streamer.js
//load modules
var ytdl = require('ytdl');
var youtube = require('youtube-feeds');
var http = require('http');
var https = require('https');

// module vars
var current_search_page=1;
var current_start_index = 1;
var videos_responses = new Array();
var yt = {};


// search videos
yt.searchVideos = function (user_search, page, filters, order, cb){
    videos_responses = new Array();
    if (page === 1){
        current_start_index = 1;
    } else {
        current_start_index = (page - 1) * 25 + 1;    
    }
    var params = {};
    if (filters === '') {
        params = {
                    'q':              ''+user_search+'',
                    'start-index' : ''+current_start_index+'',
                    'max-results':  25,
                    'orderby':        ''+order+'',
                }
    } else if (filters === 'hd') {
        params = {
                    'q':              ''+user_search+'',
                    'start-index' : ''+current_start_index+'',
                    'max-results':  25,
                    'orderby':        ''+order+'',
                    'hd' : true
                }
    } else if (filters === '3d') {
        params = {
                    'q':              ''+user_search+'',
                    'start-index' : ''+current_start_index+'',
                    'max-results':  25,
                    'orderby':        ''+order+'',
                    '3d' : true
                }
    }
    try{
        youtube.feeds.videos(params,
            function( err, data ) {
            if( err ) {
                cb(err);
            } else {
                cb(data);
            }
        });
    }catch(err){
        console.log('searchVideos err: '+err);
    }
}

// search related videos
yt.searchRelated = function (vid, page, filters, cb){
    videos_responses = new Array();
    if (page === 1){
        current_start_index = 1;
    } else {
        current_start_index = (page - 1) * 10 + 1;    
    }
    var params = {};
    if (filters === '') {
        params = {
                    'start-index' : ''+current_start_index+'',
                    'max-results':  10,
                    'orderby':        'relevance',
                }
    } else if (filters === 'hd') {
        params = {
                    'start-index' : ''+current_start_index+'',
                    'max-results':  10,
                    'orderby':        'relevance',
                    'hd' : true
                }
    } else if (filters === '3d') {
        params = {
                    'start-index' : ''+current_start_index+'',
                    'max-results':  10,
                    'orderby':        'relevance',
                    '3d' : true
                }
    }
    try{
        youtube.feeds.related(vid,params,
            function( err, data ) {
            if( err ) {
                cb(err);
            } else  {
                cb(data);
            }
        });
    }catch(err){
        console.log('searchVideos err: '+err);
    }
}

// search in categories
yt.categories = function (query, page, filters,category, cb){
    videos_responses = new Array();
    if (page === 1){
        current_start_index = 1;
    } else {
        current_start_index = (page - 1) * 25 + 1;    
    }
    var params = {};
    if (query === ''){
        var orderby = 'published';
    } else {
        var orderby = 'relevance';
    }
    if (filters === '') {
        params = {
                    'q' : ''+query+'',
                    'start-index' : ''+current_start_index+'',
                    'max-results':  25,
                    'orderby':        ''+orderby+'',
                    'category': ''+category+''
                }
    } else if (filters === 'hd') {
        params = {
                    'q' : ''+query+'',
                    'start-index' : ''+current_start_index+'',
                    'max-results':  25,
                    'orderby':        ''+orderby+'',
                    'category': ''+category+'',
                    'hd' : true
                }
    } else if (filters === '3d') {
        params = {
                    'q' : ''+query+'',
                    'start-index' : ''+current_start_index+'',
                    'max-results':  25,
                    'orderby':        'relevance',
                    'category': ''+orderby+'',
                    '3d' : true
                }
    }
    try{
        youtube.feeds.category(category,params,
            function( err, data ) {
            if( err ) {
                cb(err);
            } else  {
                cb(data);
            }
        });
    }catch(err){
        console.log('searchVideos err: '+err);
    }
}



yt.getVideoInfos = function (video_link,num,total, cb){
	if (total === 1) {
	    videos_responses = new Array();
	}
    var video = {};
    video.num = num;
    video.total = total;
    video.link = video_link;
    video.cb = cb;
    try {
        ytdl.getInfo(video.link, video, function(err, video, info){
            if(err) {
                console.log(err);
            } else {
                video.id = info.video_id;
                video.title = info.title;
                video.views = info.view_count;
                video.thumb =  info.thumbnail_url;
                video.author = info.author;
                video.duration = info.length_seconds;
                if (info.use_cipher_signature === true){
                    try{
                        var request_options = 
                        {
                            host: 'm.youtube.com',
                            headers: {'user-agent': 'Mozilla/5.0 (iPhone; U; CPU iPhone OS 4_2_1 like Mac OS X; en-us) AppleWebKit/533.17.9 (KHTML, like Gecko) Version/5.0.2 Mobile/8G4 Safari/6533.18.5'},
                            path: '/watch?ajax=1&layout=mobile&tsp=1&utcoffset=120&v='+video.id
                        }
                        var req = http.request(request_options);
                        req.on('response',function(response) { 
                            var data = new Array(); 
                            var resolutions = {};
                            response.on("data", function(chunk) {
                                data.push(chunk);
                            });
                            response.on('end',function(){
                                var d = data.join('').replace(/\)\]\}\'/g,'');
                                var datas = JSON.parse(d);
                                resolutions['480p'] = new Array();
                                try {
                                    resolutions['480p']['link'] = datas.content.video.fmt_stream_map[0].url;
                                } catch(err) {
									try { 
										resolutions['480p']['link'] = datas.content.player_data.fmt_stream_map[0].url;
									} catch (err) {
										return;
									}
                                }
                                resolutions['480p']['container'] = 'mp4';
                                video.resolutions = resolutions;
                                storeVideosInfos(video);
                            });
                        });
                        req.on('error', function(e) {
                            console.log("Got error: " + e.message);
                        });
                        req.end();
                    } catch(err){
                        console.log('youtube searchVideos err: '+err);
                    }
                } else {
                    var num=info.formats.length;
                    if ( parseInt(num) === 0) {
                            return;
                    }
                    var resolutions = {};
                    var res = [];
                    for(var i=0; i<num; i++) {
                        var resolution = info.formats[i].resolution;
                        var container = info.formats[i].container;
                        if ( in_array(resolution,res) === true || container == 'flv' || container == '3gp' ) {
                            continue;
                        }
                        res.push(resolution);
                        resolutions[resolution] = [info.formats[i].s];
                        if (info.formats[i].sig === undefined) {
							var sig = decrypt_sig(info.formats[i].s);
							var url = info.formats[i].url+'&signature='+sig+'&ratebypass=yes';
						} else {
							var url = info.formats[i].url;
						}
                            
                        resolutions[resolution]['link'] = url;
                        resolutions[resolution]['container'] = container;
                    }
                    video.resolutions = resolutions;
                    storeVideosInfos(video);
                }
            }
        });
        } catch(err) {
            console.log('getVideoInfos err: '+ err);
        }
}

//playlists
yt.searchPlaylists = function(user_search, page,cb){
    var videos_responses = new Array();
    if (page === 1){
        current_start_index = 1;
    } else {
        current_start_index = (page - 1) * 25 + 1;    
    }
    try {
        youtube.feeds.playlistSearch({
            q:              ''+user_search+'',
            'start-index' : ''+current_start_index+'',
            'max-results':  25,
            orderby:        'relevance'
            },
        function( err, data ) {
        if( err ) {
           cb(err);
        } else {
            cb(data);
        }
    });
    } catch(err) {
        console.log('searchPlaylists err: '+err);
    }   
}

//channels
yt.searchChannels = function(user_search, page,cb){
    var videos_responses = new Array();
    if (page === 1){
        current_start_index = 1;
    } else {
        current_start_index = (page - 1) * 25 + 1;    
    }
    try{
       videos_responses = new Array();
        var req=https.get('https://gdata.youtube.com/feeds/api/channels?q='+user_search+'&start-index='+current_start_index+'&v=2&alt=json&orderby=relevance');
        req.on('response',function(response) { 
            var data = new Array(); 
            response.on("data", function(chunk) {
                data.push(chunk)
            });
            response.on('end',function(){
                var datas = JSON.parse(data.join(''));
                cb(datas);
            });
        });
        req.on('error', function(e) {
            console.log("Got error: " + e.message);
        });
        req.end();
    } catch(err){
        console.log('youtube searchChannels err: '+err);
    }
}


yt.loadSongs = function(plid,length,index,cb){
    try{
        youtube.feeds.playlist(''+plid+'',{
                'start-index' : ''+index+'',
                'max-results':  25
            },
            function( err, data ) {
            if( err ) {
                cb( err );
            } else {
                cb(data, length, plid, 'youtube');
            }
        });
    } catch(err) {
        console.log('loadSongs err: '+err);
    }
}

yt.loadChannelSongs = function(url,page,cb){
    if (page === 1){
        current_start_index = 1;
    } else {
        current_start_index = (page - 1) * 25 + 1;    
    }
    try{
       videos_responses = new Array();
        var req=https.get(url+'&max-results=25&alt=jsonc&start-index='+current_start_index+'');
        req.on('response',function(response) { 
            var data = new Array(); 
            response.on("data", function(chunk) {
                data.push(chunk);
            });
            response.on('end',function(){
                var datas = JSON.parse(data.join(''));
                cb(datas);
            });
        });
        req.on('error', function(e) {
            console.log("Got error: " + e.message);
        });
        req.end();
    } catch(err){
        console.log('youtube searchChannels err: '+err);
    }
}

yt.standard = function(page,country,category,dateOrder,cb){
    if (page === 1){
        current_start_index = 1;
    } else {
        current_start_index = (page - 1) * 25 + 1;    
    }
    try{
        youtube.feeds.standard(''+country+'/'+category+'',{ 
                    'start-index' : ''+current_start_index+'',
                    'max-results':  25,
                    'time': dateOrder
            },
            function( err, datas ) {
            if( err ) {
                cb( err );
            } else {
                cb(datas);
            }
        });
    } catch(err) {
        console.log('standard feeds err: '+err);
    }
}

// store videos and return it in the right order...
function storeVideosInfos(video) {
    videos_responses[video.num]=video;
    if (videos_responses.length == video.total) {
        video.cb(videos_responses);
        videos_responses = new Array();
    }
}

// functions
function in_array (needle, haystack, argStrict) {
  var key = '',
    strict = !! argStrict;

  if (strict) {
    for (key in haystack) {
      if (haystack[key] === needle) {
        return true;
      }
    }
  } else {
    for (key in haystack) {
      if (haystack[key] == needle) {
        return true;
      }
    }
  }

  return false;
}

function swapHeadAndPosition(array, position) {
    var head  = array[0];
    var other = array[position % array.length];
    array[0] = other;
    array[position] = head;
    return array;
}

function decrypt_sig(s){
	var r;
	if (s.length == 92){
		r=s[25] + s.split('').splice(3,22) + s[0] + s.split('').splice(26,16) + s[79] + s.split('').splice(43,36) + s[91] + s.split('').splice(80,3);
	} else if (s.length == 90) {
		r=s[25] + s.split('').splice(3,22) + s[0] + s.split('').splice(26,16) + s[77] + s.split('').splice(41,36) + s[89] + s.split('').splice(78,3)
	} else if (s.length == 88) {
		r=s[48] + s.split('').splice(68,14).reverse() + s[82] + s.split('').splice(63,4).reverse() + s[85] + s.split('').splice(49,13).reverse() + s[67] + s.split('').splice(13,35).reverse() + s[3] + s.split('').splice(4,8).reverse() + s[2] + s[12];
	} else if (s.length == 87) { 
		r=s.split('').splice(4,19) + s[86] + s.split('').splice(24,61);
	} else if (s.length == 86) {
		r=s.split('').splice(2,61) + s[82] + s.split('').splice(64,18) + s[63];
	} else if (s.length == 85) {
		r=s.split('').splice(2,6) + s[0] + s.split('').splice(9,12) + s[65] + s.split('').splice(22,43) + s[84] + s.split('').splice(66,16) + s[21];
	} else if (s.length == 84) {
		r=s.split('').splice(37,47).reverse() + s[2] + s.split('').splice(27,9).reverse() + s[3] + s.split('').splice(4,22).reverse() + s[26];
	} else if (s.length == 83) {
		r=s[6] + s.split('').splice(3,3) + s[33] + s.split('').splice(7,17) + s[0] + s.split('').splice(25,8) + s[53] + s.split('').splice(34,19) + s[24] + s.split('').splice(54,34);
	} else if (s.length == 82) {
		r=s[36] + s.split('').splice(68,12).reverse() + s[81] + s.split('').splice(41,26).reverse() + s[33] + s.split('').splice(37,3).reverse() + s[40] + s[35] + s[0] + s[67] + s.split('').splice(1,32).reverse() + s[34];
	} else if (s.length == 81) {
		r=s[6] + s.split('').splice(3,3) + s[33] + s.split('').splice(7,17) + s[0] + s.split('').splice(25,8) + s[2] + s.split('').splice(34,19) + s[24] + s.split('').splice(54,27);
	} else {
		console.log('Unable to decrypt signature, key length not supported retrying might work');
		return;
	}
	var sig = r.split(',').join('');
	console.log(sig)
	return sig;
}

module.exports = yt;
