// cliphunter-js/lib/cliphunter-js.js
//load modules
var http = require('http');
var $ = require('cheerio');
// module vars
var current_search_page=1;
var current_start_index = 1;
var videos_responses = new Array();
var valid_vids = 0;
var ch = {};


// search videos
ch.searchVideos = function (query, page, duration, order, cb){
    videos_responses = new Array();
    var req=http.get('http://www.cliphunter.com/search/'+query+'/'+page+'/rel/all/all/all/900/0');
    req.on('response',function(response) { 
        var data = new Array(); 
        response.on("data", function(chunk) {
            data.push(chunk)
        });
        response.on('end',function(){
            var datas = data.join('');
            var videos = {};
            var total = $('.main_content h2',datas).text();
            var totalItems = total.match(/Found (.*?) results/)[1];
            var list=$('.tmb',datas);
            videos.totalItems = parseInt(totalItems);
            videos.total = list.length;
            videos.cb = cb;
            videos.items = [];
            for (var i=0; i<list.length; i++) {
                var infos = {};
                infos.link = 'http://www.cliphunter.com'+list[i].attribs.href;
                infos.id = infos.link.split('/')[4];
                infos.thumb = list[i].children[3].attribs.src;
                infos.time = '00:'+list[i].children[1].children[0].data.replace(/&nbsp\;/g,'');
                infos.title = list[i].attribs.alt;
                infos.views = '...';
                infos.author = 'unknown';
                getVideoInfos(videos,infos,i);
            }
        });
    });
    req.on('error', function(e) {
        console.log("Got error: " + e.message);
    });
    req.end();
}

function decrypt(f){var c="mbcidgsaztuynlfvjxrqpewhko";var e="";var d=/[a-z]/;for(var b=0;b<f.length;b++){var a=f.charAt(b);if(d.test(a)){e+=String.fromCharCode(c.indexOf(a)+97)}else{if(a=="$"){e+=":"}else{if(a=="="){e+="/"}else{if(a=="&"){e+="."}else{if(a=="^"){e+="&"}else{if(a=="("){e+="="}else{e+=f.charAt(b)}}}}}}}return e}

function getVideoInfos(videos,infos,num) {
    var req=http.get('http://www.cliphunter.com/a/dwnlinks/'+infos.id+'');
    var data = new Array(); 
    req.on('response',function(response) { 
        response.on("data", function(chunk) {
            data.push(chunk);
        });
        response.on('end',function(){
            var datas = data.join('');
            $('.dwnlink',datas).each(function(i,elem) {
                var cont = $(this).parent().prev().prev().prev().prev()[0].children[0].data;
                if (cont === 'MP4') {
                    var resol = $(this).parent().prev().prev().text().split('x')[0].trim();
                    if (parseInt(resol) === 640) {
                        var res = '480p'; 
                    } else {
                        var res = '360p';
                    }
                    var link = decrypt($(this)[0].attribs.href);
                    infos.resolutions=[];
                    infos.resolutions[res] = {};
                    infos.resolutions[res]['link'] = link;
                    infos.resolutions[res]['container'] = 'mp4';
                }
            });
            storeVideosInfos(videos,infos,num);
        });
    });
    req.on('error', function(e) {
        console.log("Got error: " + e.message);
    });
    req.end();
}

// search in categories
ch.categories = function (query, page, filters,category, cb){
    videos_responses = new Array();
    if (page === 1){
        current_start_index = 1;
    } else {
        current_start_index = page * 25 + 1;    
    }
    var params = {};
    if (query === ''){
        var orderby = 'published';
    } else {
        var orderby = 'relevance';
    }
    if (filters === '') {
        params = {
                    'q' : ''+query+'',
                    'start-index' : ''+current_start_index+'',
                    'max-results':  25,
                    'orderby':        ''+orderby+'',
                    'category': ''+category+''
                }
    } else if (filters === 'hd') {
        params = {
                    'q' : ''+query+'',
                    'start-index' : ''+current_start_index+'',
                    'max-results':  25,
                    'orderby':        ''+orderby+'',
                    'category': ''+category+'',
                    'hd' : true
                }
    } else if (filters === '3d') {
        params = {
                    'q' : ''+query+'',
                    'start-index' : ''+current_start_index+'',
                    'max-results':  25,
                    'orderby':        'relevance',
                    'category': ''+orderby+'',
                    '3d' : true
                }
    }
    try{
        youtube.feeds.category(category,params,
            function( err, data ) {
            if( err ) {
                cb(err);
            } else  {
                cb(data);
            }
        });
    }catch(err){
        console.log('searchVideos err: '+err);
    }
}

ch.getVideoInfos = function (video_link,num,total, cb){
	if (total === 1) {
	    videos_responses = new Array();
	}
    var video = {};
    video.num = num;
    video.total = total;
    video.link = video_link;
    video.cb = cb;
    try {
        ytdl.getInfo(video.link, video, function(err, video, info){
            if(err) {
                console.log(err);
            } else {
                video.id = info.video_id;
                video.title = info.title;
                video.views = info.view_count;
                video.thumb =  info.thumbnail_url;
                video.author = info.author;
                video.duration = info.length_seconds;
                if (info.use_cipher_signature === true){
                    try{
                        var request_options = 
                        {
                            host: 'm.youtube.com',
                            headers: {'user-agent': 'Mozilla/5.0 (iPhone; U; CPU iPhone OS 4_2_1 like Mac OS X; en-us) AppleWebKit/533.17.9 (KHTML, like Gecko) Version/5.0.2 Mobile/8G4 Safari/6533.18.5'},
                            path: '/watch?ajax=1&layout=mobile&tsp=1&utcoffset=120&v='+video.id
                        }
                        var req = http.request(request_options);
                        req.on('response',function(response) { 
                            var data = new Array(); 
                            var resolutions = {};
                            response.on("data", function(chunk) {
                                data.push(chunk);
                            });
                            response.on('end',function(){
                                var d = data.join('');
                                var datas = JSON.parse(d.replace(/\)\]\}\'/,''));
                                resolutions['480p'] = new Array();
                                try {
                                    resolutions['480p']['link'] = datas.content.video.fmt_stream_map[0].url;
                                } catch(err) {
                                    return;
                                }
                                resolutions['480p']['container'] = 'mp4';
                                video.resolutions = resolutions;
                                storeVideosInfos(video);
                            });
                        });
                        req.on('error', function(e) {
                            console.log("Got error: " + e.message);
                        });
                        req.end();
                    } catch(err){
                        console.log('youtube searchVideos err: '+err);
                    }
                } else {
                    var num=info.formats.length;
                    if ( parseInt(num) === 0) {
                            return;
                    }
                    var resolutions = {};
                    var res = [];
                    for(var i=0; i<num; i++) {
                        var resolution = info.formats[i].resolution;
                        var container = info.formats[i].container;
                        if ( in_array(resolution,res) === true || container == 'flv' || container == '3gp' ) {
                            continue;
                        }
                        res.push(resolution);
                        resolutions[resolution] = [];
                        var url = info.formats[i].url;
                            
                        resolutions[resolution]['link'] = url;
                        resolutions[resolution]['container'] = container;
                    }
                    video.resolutions = resolutions;
                    storeVideosInfos(video);
                }
            }
        });
        } catch(err) {
            console.log('getVideoInfos err: '+ err);
        }
}


// store videos and return it in the right order...
function storeVideosInfos(video,infos,num) {
    video.items.push(infos); 
    videos_responses[num]=video;
    if (videos_responses.length == video.total) {
        video.cb(videos_responses);
        videos_responses = new Array();
    }
}

// functions
function in_array (needle, haystack, argStrict) {
  var key = '',
    strict = !! argStrict;

  if (strict) {
    for (key in haystack) {
      if (haystack[key] === needle) {
        return true;
      }
    }
  } else {
    for (key in haystack) {
      if (haystack[key] == needle) {
        return true;
      }
    }
  }

  return false;
}

function swapHeadAndPosition(array, position) {
    var head  = array[0];
    var other = array[position % array.length];
    array[0] = other;
    array[position] = head;
    return array;
}

module.exports = ch;
