// superhqporn-js/lib/superhqporn-js.js
//load modules
var http = require('http');
var $ = require('cheerio');
// module vars
var current_search_page=1;
var current_start_index = 1;
var videos_responses = new Array();
var valid_vids = 0;
var shq = {};


// search videos
shq.searchVideos = function (query, page, duration, order, cb){
    videos_responses = new Array();
    var req=http.get('http://www.superhqporn.com/?s=search&search='+query+'&p='+page+'');
    req.on('response',function(response) { 
        var data = new Array(); 
        response.on("data", function(chunk) {
            data.push(chunk);
        });
        response.on('end',function(){
            var datas = $(data.join(''));
            var videos = {};
            var totalItems = $('.header',datas).text().match(/ (.*) videos/)[1];
            var list=$('.thumb',datas);
            videos.totalItems = parseInt(totalItems);
            videos.total = list.length;
            videos.cb = cb;
            videos.items = [];
            for (var i=0; i<list.length; i++) {
                var infos = {};
                infos.link = 'http://www.superhqporn.com'+list[i].children[0].children[0].attribs.href;
                infos.id = infos.link.match(/v=(.*)/)[1];
                infos.thumb = list[i].children[0].children[0].children[0].attribs.src;
                infos.time = '00:'+list[i].children[1].children[0].data;
                infos.title = list[i].children[0].children[0].attribs.title;
                getVideoInfos(videos,infos,i);
            }
        });
    });
    req.on('error', function(e) {
        console.log("Got error: " + e.message);
    });
    req.end();
}

function getVideoInfos(videos,infos,num) {
    var req=http.get('http://www.superhqporn.com/?v='+infos.id+'');
    var data = new Array(); 
    req.on('response',function(response) { 
        response.on("data", function(chunk) {
            data.push(chunk);
        });
        response.on('end',function(){
            var datas = data.join('');
            infos.views = '...';
            infos.author = 'unknown';
            var link = datas.match('url: \'(.*?)\'')[1];
            infos.resolutions=[];
            infos.resolutions['480p'] = {};
            infos.resolutions['480p']['link'] = link;
            infos.resolutions['480p']['container'] = 'mp4';
            storeVideosInfos(videos,infos,num);
        });
    });
    req.on('error', function(e) {
        console.log("Got error: " + e.message);
    });
    req.end();
}



// store videos and return it in the right order...
function storeVideosInfos(video,infos,num) {
    video.items.push(infos); 
    videos_responses[num]=video;
    if (videos_responses.length == video.total) {
        video.cb(videos_responses);
        videos_responses = new Array();
    }
}

// functions
function in_array (needle, haystack, argStrict) {
  var key = '',
    strict = !! argStrict;

  if (strict) {
    for (key in haystack) {
      if (haystack[key] === needle) {
        return true;
      }
    }
  } else {
    for (key in haystack) {
      if (haystack[key] == needle) {
        return true;
      }
    }
  }

  return false;
}

function swapHeadAndPosition(array, position) {
    var head  = array[0];
    var other = array[position % array.length];
    array[0] = other;
    array[position] = head;
    return array;
}

module.exports = shq;
